internal/controlplane/common.go

```diff
@@ -59,3 +67,32 @@ func getProjectFromRequestOrDefault(ctx context.Context, in ProjectIDGetter) (uu
 	}
 	return parsedProjectID, nil
 }
+
+func getProviderFromRequestOrDefault(
+	ctx context.Context,
+	store db.Store,
+	in ProviderNameGetter,
+	projectId uuid.UUID,
+) (db.Provider, error) {
+	providers, err := store.ListProvidersByProjectID(ctx, projectId)
+	if err != nil {
+		return db.Provider{}, status.Errorf(codes.InvalidArgument, "cannot retrieve providers: %s", err)
+	}
+	// if we do not have a provider name, check if we can infer it
+	if in.GetProvider() == "" {
+		if len(providers) == 1 {
+			return providers[0], nil
+		}
+		return db.Provider{}, status.Errorf(codes.InvalidArgument, "cannot infer provider name: %s", err)
```

Comment:
Do you want to include `len(providers)` in the error?  Also, I think `err` is `nil` here, so you probably don't want to return that.

Consider using `util.UservisibleError` so that the error gets back to the CLI.

```suggestion
		return db.Provider{}, util.UserVisibleError(codes.InvalidArgument, "cannot infer provider, there are %d providers available", len(providers))
```